predicators/approaches/delivery_approach.py

```diff
@@ -0,0 +1,197 @@
+                                                self.pred_is_home_base,
+                                                init_ground_atoms)
+        obj_home_base = next(iter(obj_home_bases), None)
+
+        # Extracting goal objects
+        goal_at_loc = self._extract_at_loc_optional(obj_locs, task.goal)
+        goal_wants_paper = filter_predicated_objs(obj_locs,
+                                                  self.pred_wants_paper,
+                                                  task.goal)
+        goal_satisfied = filter_predicated_objs(obj_locs, self.pred_satisfied,
+                                                task.goal)
+        goal_unpacked = filter_predicated_objs(obj_papers, self.pred_unpacked,
+                                               task.goal)
+        goal_carrying = filter_predicated_objs(obj_papers, self.pred_carrying,
+                                               task.goal)
+
+        # Joint object extraction
+        obj_available_papers = obj_papers - goal_unpacked
+        obj_satisfying_papers = obj_available_papers - goal_carrying
+
+        # Sanity checks (satisfies all the "safe" and "home_base" goals)
+        if not filter_predicated_objs(obj_locs, self.pred_safe,
+                                      task.goal) <= obj_safe_locs:
+            raise ApproachFailure("Cannot make new safe locations")
+        if not filter_predicated_objs(obj_locs, self.pred_is_home_base,
+                                      task.goal) <= obj_home_bases:
+            raise ApproachFailure("Cannot create new bases")
+
+        def _policy(state: State) -> Action:
+            state = cast(_PDDLEnvState, state)
+            ground_atoms = state.get_ground_atoms()
+
+            obj_at_loc = self._extract_at_loc(obj_locs, ground_atoms)
+            obj_carried_papers = filter_predicated_objs(
+                obj_papers, self.pred_carrying, ground_atoms)
+            obj_locs_to_satisfy = goal_satisfied - filter_predicated_objs(
+                obj_locs, self.pred_satisfied, ground_atoms)
+
+            # Pick up enough satisfying papers
+            if len(obj_carried_papers) < len(obj_locs_to_satisfy):
+                if obj_home_base is None:
+                    raise ApproachFailure("Need a home base")
+                if obj_at_loc != obj_home_base:
+                    return self._move(state, obj_at_loc, obj_home_base)
+
+                obj_satisfying_paper = next(
+                    iter(obj_satisfying_papers - obj_carried_papers), None)
+                if obj_satisfying_paper is None:
+                    raise ApproachFailure(
+                        "Too little paper that can satisfy locations")
+                return self._pick_up(state, obj_satisfying_paper,
+                                     obj_home_base)
+
+            # If already at a location to satisfy, satisfy it
+            if obj_at_loc in obj_locs_to_satisfy:
+                carried_paper = obj_carried_papers.pop()
```

Comment:
another very minor thing: we generally try to make sure that our code is fully deterministic so that we can more easily reproduce and debug things. popping from a set in Python is not necessarily deterministic, so we would prefer to do this another way (e.g., maintaining a list instead of a set)